### DISCUSSION TITLE
Way to determine object type in `.cue`

### DISCUSSION THREAD

**Discussion by vnzongzna** [cuelang/cue#518] [General]
I need to determine if a value will of type `struct` or `list` to make the decision on adding new fields in it.
For example
```
sample: [...] | {...}
sample: batman: is: "back"
if sample == [...]  {
   array: "yes" 
}
if sample == {...} {
   struct: "yes" 
}
```

**Comment by vnzongzna** [comment]
I got my solution [here in slack](https://cuelang.slack.com/archives/CLT3ULF6C/p1576619870062700). which pointed to checking against bottom. So a possible solution is 
```
sample: [...] | {...}
sample: batman: is: "back"
if ( sample & [...]  )!= _|_ {
   array: "yes" 
}
if ( sample & {...}  )!= _|_ {
   struct: "yes" 
}
```
While this solution does work would it make sense to have an inbuilt function to return them?

**Reply by vnzongzna** [reply]
Because as the author of the whole config I do know the possible values `sample` will have but maybe in a big enough configuration getting help from build-in will be faster.

**Comment by mpvl** [comment]
> While this solution does work would it make sense to have an inbuilt function to return them?

I think it may make sense to have a builtin for that. A more general builtin to make this nicer could be something like `defined(sample&[...])` or even `subsumes([...], sample)`. Both would be more consistent with values being types, and provide provide a larger range of possibilities.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/518.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
